### PR TITLE
[SPARK-13478] [yarn] Use real user when fetching delegation tokens.

### PR DESCRIPTION
**Pull request by vanzin**
The Hive client library is not smart enough to notice that the current
user is a proxy user; so when using a proxy user, it fails to fetch
delegation tokens from the metastore because of a missing kerberos
TGT for the current user.

To fix it, just run the code that fetches the delegation token as the
real logged in user.

Tested on a kerberos cluster both submitting normally and with a proxy
user; Hive and HBase tokens are retrieved correctly in both cases.

### CODE REVIEW

**Comment by harishreedharan** [comment]
So I have not tested using the keytab-based login with proxy user stuff at all. We get delegation tokens even there - does this issue affect that as well?

**Comment by harishreedharan** [comment]
This looks like it might affect HDFS tokens as well and error that looks like this might come up during the initial token renewal:
```
WARN UserGroupInformation: PriviledgedActionException as:hari (auth:PROXY) via hdfs@EXAMPLE (auth:KERBEROS) cause:org.apache.hadoop.security.AccessControlException: hari tries to renew a token with renewer hdfs
```

In addition to the code that gets the new tokens, I think the [`getTokenRenewalInterval`](https://github.com/apache/spark/blob/master/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala#L580) method also needs to be be run as the real user.

**Comment by vanzin** [comment]
> keytab-based login with proxy user stuff at all

That is not expected to work. Just use the proper user's credentials if you're going to provide a principal and a keytab.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 1]
```diff
@@ -171,8 +173,8 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
    * @param username the username of the principal requesting the delegating token.
    * @return a delegation token
    */
-  private[yarn] def obtainTokenForHiveMetastoreInner(conf: Configuration,
-      username: String): Option[Token[DelegationTokenIdentifier]] = {
+  private[yarn] def obtainTokenForHiveMetastoreInner(conf: Configuration):
```
Comment (tgravescs) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: only reason I thought this was passed in was for testing, is that not the case?

**Comment by tgravescs** [comment]
the code looks fine to me.  Do we need any more documentation to explain proxy user vs keytab?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala (hunk 1 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: The sole test for this code didn't seem to do any verification of the given user name, so it didn't seem necessary to have a parameter here.

**Comment by vanzin** [comment]
> Do we need any more documentation to explain proxy user vs keytab?

Ok, ok, I'll add something.

**Comment by tgravescs** [comment]
looks fine.  +1 pending jenkins.  @harishreedharan any other concerns?

**Comment by harishreedharan** [comment]
LGTM

**Comment by vanzin** [comment]
Merging this.

**Comment by WangTaoTheTonic** [comment]
hi @vanzin, how about spark sql in this issue, in your view? as in spark sql it will revoke SessionState.start in which will finally connected to metastore in Hive.

**Comment by vanzin** [comment]
@WangTaoTheTonic please don't comment on closed PRs. If there's a bug in Spark SQL, file a bug.

**Pull request closed without merging**